### ISSUE TITLE
Migrate Level1 and Bluestar/Level0 develop to VS 2015

### ISSUE THREAD

**Issue by dan31** [Starcounter/Starcounter#2981]
To move adoption of new kernel to 2.x, develop Level1 has to be moved to vs-2015 (classical, non-dnx project types C#/F# + native projects C++). The opposite direction (new kernel to vs-2012) is no longer viable. @alemoi is busy, and I call for @per-samuelsson or @chrhol to do the move. It is not trivial, and perhaps will take a couple of days. I can assist with native parts on Monday-Tuesday. I'm thinking @chrhol is more suitable for the work? Guys? This must be done high prio. CC @alemoi @eriohl

**Comment by per-samuelsson** [comment]
I'd be really glad if this was not on me. :cry: I've just recollected a list of issues I plan todo around process management, server refining and the developer experience in general, based on the new directions from yesterday. Plus I totally suck on Visual Studio / msbuild / projects, etc.

If it turns out that really no one else can take this, I'll still do it o/c, but I'll need a few paid Mallorca days with free drinks in return! :cocktail: :sunny: :sunrise:

**Comment by chrhol** [comment]
I can take it, I dont have anything super urgent at the moment assigned to me and need to wait for @alemoi to catch up on some fixes I need for the other stuff I'm working on.

**Comment by per-samuelsson** [comment]
> chrhol self-assigned this 41 seconds ago

Next time we play poker, I'll fold the nuts to you heads up on the river! :moneybag:

**Comment by chrhol** [comment]
I guess this also includes, besides simply migrating Level1, to update/make sure that the buildservers and downloadserver correctly builds and generates downloads.

**Comment by eriohl** [comment]
And level0/develop needs to updated also I guess.

**Comment by eriohl** [comment]
Alternative is to look into moving level0 back to vs2012. Looked (very) briefly at this yesterday and concluded that a lot of the NewSQL stuff won't compile.  Otherwise this should not be much of an issue I think. And we don't actually need these. So one option could be to disable them when building with 2012 (possible to to in the cmake file maybe).

But on the other hand: Since 2.x will get an extended lease on life now it will have to be done eventually anyway. So...

**Comment by chrhol** [comment]
Alright, then we have two options to move forward. 

To me it sounds better to migrate to VS 2015, even if it's a bit more work since we probably want to do that anyway later but on the other hand we will most likely save some time going the other direction. @dan31 WDYT?

**Comment by chrhol** [comment]
The decision is to migrate to VS2015. I will keep everything in this issue even though there are parts outside of Level1 that needs to be migrated. 

Steps:
- Migrate Level0/Bluestar develop
- Migrate Level1 develop
- Update buildservers 
- Update downloadserver

**Comment by chrhol** [comment]
Just to be clear so I don't do anything stupid, migrate Level0 develop to VS2015 means that we switch over to the compiler and build tools for 2015 or should we keep the current?

**Comment by per-samuelsson** [comment]
> I will keep everything in this issue even though there are parts outside of Level1 that needs to be migrated.

Sounds reasonable, but lets change the title.

**Comment by dan31** [comment]
@chrhol, did you conclude with @eriohl that Leve0/Bluestar develop as it is there today has to be adopted to vs-2015? Or instead, pnext kernel will be reused in a form of cmake-generated project for vs-2015 or compiled binary? (there are also cross-dependencies between Level0 and Level1).

**Comment by chrhol** [comment]
@dan31, 
I'm not really sure what you mean by if pnext should be reused. Do you mean that we should switch over to pnext now? (pnext is not the new kernel?)

I talked with @eriohl yesterday and he didn't think it would be that difficult to migrate Level0 develop.  And that together with that we don't have to maintain two branches (again) while waiting for the new kernel sounds like a good way to go. Also we will have time to fix any eventual problems that arises (buildservers, tests, downloadserver) and not have everything at once if we do the switch when the new kernel is ready.

**Comment by dan31** [comment]
> I'm not really sure what you mean by if pnext should be reused. Do you mean that we should switch over to pnext now? (pnext is not the new kernel?)

Yeah, by pnext I meant the new kernel. But I got your scheme now, first you migrate all develop to vs-2015, including level0 develop, that's the first result and it becomes develop. Then @eriohl plugs in the new kernel in some branch of that result. This branch is short-living enough and it eventually becomes develop as well. So we have short-living branches all the way. I assume you agreed upon things in the best way, I was clarifying for myself, thanks for explaining! The challenge here is that new kernel is cross-platform and thus is built based on CMake. So, when new kernel is in place in develop, it isn't formed by a pure vs-2015 projects, right? There seems to be some hybrid, when, to build develop, you'd have to build Blue from CMakeLists first, then build Level1 natively as vs-2015 solution.

**Comment by chrhol** [comment]
I have migrated both Level0 develop and Level1 develop locally. I tested compiling, installing and running some of our apps and so far everything seems to be working fine. Boost and gtest, if you build level 0, have to be updated (compiled with vs2015). Both packages can be found on the buildserver-ftp

Boost generates quite a lot of warnings so I had to disable 'treat warnings as errors' in the projects that use boost.

I've pushed both branches and will keep them up-to-date with develop until we decide to do the actual move when everything is in order. I named the branches the same in both Level1 and Level0, `issue\2981`

Some additional things:
- The default settings when installing VS2015 does NOT include C++ and F# so these have to be added when choosing what to install
- The version of DotCover that we use does not support VS2015 and our license is not valid for upgrading so we will need to buy new licenses for DotCover 3.0 if we want to continue to use it.

I will continue with investigatíng what needs to be changed/installed on the buildservers in order to make them work.

**Comment by per-samuelsson** [comment]
> I have migrated both Level0 develop and Level1 develop locally. I tested compiling, installing and running some of our apps and so far everything seems to be working fine.

Wow, sweet. :ghost:

**Comment by dan31** [comment]
Great @chrhol! :+1: Is that the latest versions of boost & gtest that produce walls of warnings? I think we should deliberately disable the warnings with pragmas and return the strict warning check (CC @chrhol @eriohl). If I recall correctly, that way it has been done already in some projects?

**Comment by eriohl** [comment]
Boost warnings I probably already addressed this way in level1 pnext (chunks). Boost not yet working properly with 2015 last checked.

Unsure what the GTest stuff is about but we use this fine in pnext. Although maybe with disabled warnings.

**Comment by chrhol** [comment]
> Unsure what the GTest stuff is about but we use this fine in pnext. Although maybe with disabled warnings.

No problems with GTest I think, warnings comes from boost and ipc (maybe not all warnings comes from boost). 

I will spend some time going through them, after I get the buildserver to work.

**Comment by eriohl** [comment]
Merged 2981 into pnext.

**Comment by chrhol** [comment]
I think I have gotten as far as I can with this. Most of the warnings are fixed and 'treat warnings as errors' is set back to true on the projects I changed before. The buildservers have been updated.

What is left, is to fix so that the needed c++ runtime is included and installed from the installer and upgrade the downloadserver.

I will assign the issue to @alemoi since he will fix the c++ runtime in the installer.

**Steps for building with 2015**
- Install 2015, choose Custom installation, select options `Visual C++`, `Visual F#` and `Visual Studio Extensibility Tools`
- Unpack `\\scbuildserver\FTP\SCDev\ThirdParty\boost_1_58_0_vc140.zip` (or .7z) to `c:` (the paths should be `c:\boost\include` and `c:\boost\lib`)
- If building Level0: Unpack `\\scbuildserver\FTP\SCDev\ThirdParty\gtest_vc140.zip` to `c:` (the paths should be `c:\gtest\vc140\build` and `c:\gtest\vc140\include`)
- If you want to build before the merge check out the branch `issue\2981` in both Level1 and Level0. 
- If you want to run in the buildsystem set branch-parameters for Level1, Level0 and BuildSystem to `issue\2981`

**Comment by dan31** [comment]
One more thing there, popped up while discussing with @alemoi today. Should it be the case that we all switch to .NET 4.6 there? Including: building all managed projects, and Starcounter templates for VS? I personally don't see practical reason to run 4.5/4.5.1 instead of 4.6. CC @per-samuelsson @k-rus @alemoi @eriohl @chrhol.

**Comment by per-samuelsson** [comment]
> Should it be the case that we all switch to .NET 4.6 there?

Sounds reasonable. And if there are problems building our stuff, we can just back it down to 4.5 again, shouldn't be to much work.

And then the road is open to C# 6.0! :)

**Comment by dan31** [comment]
@per-samuelsson Exactly, and this is so cool :bomb: I was just dazzling @alemoi with Elvis operator https://msdn.microsoft.com/en-us/library/dn986595.aspx today :)

**Comment by alemoi** [comment]
And I was totally dazzled! telling you! :)

**Comment by bigwad** [comment]
Actually, AFAIK, you don't have to target .Net 4.6 to use c# 6. Compiling with vs 2015 is enough. So don't stop yourself :)

**Comment by chrhol** [comment]
@k-rus found some problems with nuget today. It seems that for the newer versions of nuget you no longer need the `.nuget` folder and `nuget.targets` file. I have removed all references from all projects in Level1. 

For the buildsystem I added scripts to download latest `nuget.exe` and execute a restore before building. This means that when we do the switch we can remove `\\scbuildserver\FTP\SCDev\ThirdParty\packages\` since they will be downloaded.

**Comment by alemoi** [comment]
That would super awesome regarding auto pull of needed versions of Nuget packages! Did you test this on automatic builds on buildservers? Great if it works!

**Comment by chrhol** [comment]
> That would super awesome regarding auto pull of needed versions of Nuget packages! Did you test this on automatic builds on buildservers? Great if it works!

Yeah, I've changed it in the buildsystem-repo (`issue/2981`), and it seems to be working without problem. I will show you what I changed when you are in the office. Maybe we want to refine it a bit.

**Comment by k-rus** [comment]
Another finding is that VS projects cannot be built on a clean machine with VS2015 without previous versions of VS.

**Comment by chrhol** [comment]
Everything should be fixed now for the issues that have been found.

And we have merged everything to develop!

**Comment by per-samuelsson** [comment]
> @per-samuelsson Exactly, and this is so cool :bomb: I was just dazzling @alemoi with Elvis operator https://msdn.microsoft.com/en-us/library/dn986595.aspx today :)

Yeah, that one is nice for sure. I also like the neat new `nameof` operator. That one I've longed for for years, since it makes refactorings much more accurate and saves us from quite a few problems we have today.

I also like the new auto-property initializers, while I don't particulary like the new lamba-way of declaring method bodies. Methods are the holy grail for me, you don't mess with them like that. :cold_sweat:

**Comment by per-samuelsson** [comment]
> **Steps for building with 2015** [...]

Vedict from 9km south of HQ: Worked like a charm, straight out of the box! :clap: to @chrhol and @alemoi.

**Comment by alemoi** [comment]
In installer, after research and tests, I had to embed VS CRTs setups, so the size of installer is now increased significantly :(
CRTs are detected/installed as a first thing in installer now (e.g. tested with clean Win7, works fine) before the .NET framework (which is still 4.5 until we decide to switch).

**Comment by dan31** [comment]
> In installer, after research and tests, I had to embed VS CRTs setups, so the size of installer is now increased significantly :( CRTs are detected/installed as a first thing in installer now (e.g. tested with clean Win7, works fine) before the .NET framework (which is still 4.5 until we decide to switch).

Unfortunately this is as good as we get it today. Redistributing with self-instanller is a Microsoft recommended way, and we pay a price of size for simplicity of installation.

**Comment by eriohl** [comment]
Not sure if the size is actually a problem. But could we not have the installer download the CRT and CLR setups in case they are not installed (which the probably are on a developer machine).

**Comment by per-samuelsson** [comment]
> But could we not have the installer download the CRT and CLR setups in case they are not installed (which the probably are on a developer machine).

+1 on this. Not sure if it applies to this case (I guess @alemoi  knows that), but in general, every opportunity we have for on-demand dependency resolving, we should take.

**Comment by dan31** [comment]
Yeah @alemoi has chosen the easiest way to support so far. @alemoi, do you think there is any complication for doing @eriohl's suggestion (it isn't high priority, though necessary soon)? I suggest there should be no obstacle to keep it silent in downloadable scenario as well.

**Comment by alemoi** [comment]
Yes, I did the easiest way so far. But downloading is fine as long as links for CRTs are persistent.

**Comment by dan31** [comment]
> Yes, I did the easiest way so far. But downloading is fine as long as links for CRTs are persistent.

I would suggest (didn't check it though) that MS provides some special API/REST API for that. Remember, link to a download page they always give persistent, but when you click download, they generate you some hash-like address. Note: license doesn't allow you to store those installers elsewhere rather than taking them from MS website. It'd been the case 6 years ago, presumably the same applies as of today.

**Comment by chrhol** [comment]
We have had some problems with the VS extension after the upgrade (#3077). Hopefully the problems will be solved now. It boils down to this: you can only reference VS-SDK dlls that are from the lowest version you wish to support. Since we support extension for 2012, 2013 and 2015 this means that we can only use VS2012 SDK dlls.

The problem with this is that the SDK for 2012 is not included in the VS2015 SDK, so it's not enough to just install the SDK for VS2015. AND you are not allowed to just install VS2012 SDK without having VS2012 installed.

I found a package on nuget that contains all dll's we need (https://www.nuget.org/packages/VSSDK.Shell.11/), but this is not an offical release, so I don't know how dependent we want to be on this. 

Basically we have two options: 
1. Use the nuget package. Extension will build with just installing VS2015.
2. Install VS2012 and 2012 SDK together with 2015.

These are the only solutions I found. Currently option 1. is pushed to develop.

**Comment by alemoi** [comment]
I guess this one can be closed now? Since @chrhol resolved all the problems we had.

**Comment by chrhol** [comment]
> I guess this one can be closed now? Since @chrhol resolved all the problems we had.

Yeah, you are right. I will close it!
